modules/salesforce_jwt/src/Plugin/SalesforceAuthProvider/SalesforceJWTPlugin.php

```diff
@@ -0,0 +1,256 @@
+    if (!$this->keyRepository()) {
+      $form_state->setError($form, $this->t('JWT Auth requires <a href="https://drupal.org/project/key">Key</a> module. Please install before adding a JWT Auth config.'));
+      return;
+    }
+
+    parent::validateConfigurationForm($form, $form_state);
+    try {
+      $settings = $form_state->getValue('provider_settings');
+      $this->getToken($settings['login_url']);
+      \Drupal::messenger()->addStatus(t('Successfully connected to Salesforce as user %name.', ['%name' => $this->getIdentity()['display_name']]));
+    }
+    catch (\Exception $e) {
+      $form_state->setError($form, $this->t('Failed to connect to Salesforce: %message', ['%message' => $e->getMessage()]));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    parent::submitConfigurationForm($form, $form_state);
+    $this->setConfiguration($form_state->getValues());
+  }
+
+  /**
+   * Gets a token from the given JWT OAuth endpoint.
+   */
+  protected function getToken($login_url) {
+    // Initialize access token.
+    $assertion = $this->generateAssertion();
+    $data = [
+      'grant_type' => 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+      'assertion' => $assertion,
+    ];
+    $response = $this->httpClient->retrieveResponse(new Uri($login_url . static::AUTH_TOKEN_PATH), $data, ['Content-Type' => 'application/x-www-form-urlencoded']);
+    $token = $this->parseAccessTokenResponse($response);
+    $this->storage->storeAccessToken($this->service(), $token);
+
+    // Initialize identity.
+    $headers = [
+      'Authorization' => 'OAuth ' . $token->getAccessToken(),
+      'Content-type' => 'application/json',
+    ];
+    $data = $token->getExtraParams();
+    $response = $this->httpClient->retrieveResponse(new Uri($data['id']), [], $headers);
+    $identity = $this->parseIdentityResponse($response);
+    $this->storage->storeIdentity($this->service(), $identity);
+
+    return $token;
+  }
+
+  /**
+   * Refreshes an OAuth2 access token.
+   *
+   * @param TokenInterface $token
+   *
+   * @return TokenInterface $token
+   *
+   * @throws \OAuth\OAuth2\Service\Exception\MissingRefreshTokenException
+   * @throws \OAuth\Common\Http\Exception\TokenResponseException
+   */
+  public function refreshAccessToken(TokenInterface $token) {
+    $token = $this->getToken($this->getLoginUrl());
+    return $token;
+  }
+
+  /**
+   * Key repository wrapper.
+   *
+   * @return \Drupal\key\KeyRepository|FALSE
+   *   The key repo.
+   */
+  protected function keyRepository() {
+    if (!\Drupal::hasService('key.repository')) {
+      return FALSE;
+    }
+    return \Drupal::service('key.repository');
+  }
+
+  /**
+   * Returns a JWT Assertion to authenticate.
+   *
+   * @return string
+   *   JWT Assertion.
+   */
+  private function generateAssertion() {
```

Comment:
I wonder if we should use https://github.com/firebase/php-jwt for this instead of rolling our own.

Reply:
Sure, why not. Eliminated a dozen or so lines of code.